### PR TITLE
Fallback to current skin combo colours if none provided on beatmap skin

### PR DESCRIPTION
**Pull request by frenzibyte**
- Fixes using beatmap skin with no combo colours provides defualt combo colours instead of current skin's.

- To reproduce this issue on current master:

1. Select a skin with custom combo colours.
2. Play a beatmap that doesn't provide combo colours with `beatmap skin` enabled.
3. Result: Combo colours shown are of the default skin's combo colours not the current skin's custom combo colours.

### CODE REVIEW

File: osu.Game.Tests/Resources/skin-empty.ini [hunk 1]
```diff
@@ -1,2 +0,0 @@
-[General]
-Name: test skin
```
Comment (peppy) [inline osu.Game.Tests/Resources/skin-empty.ini]: what's the reason for removing this?

Comment (iiSaLMaN) [reply osu.Game.Tests/Resources/skin-empty.ini]: Doesn't seem to be used, although i've realized it has a usage with something i'm working on, will return.

File: osu.Game.Tests/Skins/LegacySkinDecoderTest.cs [hunk 2]
```diff
@@ -13,28 +13,23 @@ namespace osu.Game.Tests.Skins
     [TestFixture]
     public class LegacySkinDecoderTest
     {
-        [TestCase(true)]
-        [TestCase(false)]
-        public void TestDecodeSkinColours(bool hasColours)
```
Comment (peppy) [inline osu.Game.Tests/Skins/LegacySkinDecoderTest.cs]: why did you remove the test for no colours?

File: osu.Game/Skinning/LegacyBeatmapSkin.cs [hunk 3]
```diff
@@ -9,6 +9,9 @@ namespace osu.Game.Skinning
 {
     public class LegacyBeatmapSkin : LegacySkin
     {
+        // Null should be returned in the case of no colours provided to fallback into current skin's colours.
```
Comment (peppy) [inline osu.Game/Skinning/LegacyBeatmapSkin.cs]: this comment is no sense. null what? no colours provided to fallback?

File: osu.Game/Skinning/LegacySkin.cs [hunk 4]
```diff
@@ -63,7 +68,12 @@ protected override void Dispose(bool isDisposing)
                     switch (global)
                     {
                         case GlobalSkinConfiguration.ComboColours:
-                            return SkinUtils.As<TValue>(new Bindable<List<Color4>>(Configuration.ComboColours));
+                            if (Configuration.ComboColours.Any())
```
Comment (peppy) [inline osu.Game/Skinning/LegacySkin.cs]: you've bypassed `Configuration`. configuration is public. now you can get two values for combo colours depending on how you access.

File: osu.Game/Skinning/LegacyBeatmapSkin.cs (hunk 3 above)
Comment (iiSaLMaN) [reply osu.Game/Skinning/LegacyBeatmapSkin.cs]: Was afraid of that (my explanation is shit). null should be returned if the beatmap skin doesn't provide any combo colours when trying to retrieve with `GetConfig(...)`, this will make `SkinProvidingContainer` retrieve from the current skin's (`SkinProvidingContainer.fallbackSource`) combo colours.

File: osu.Game.Tests/Skins/LegacySkinDecoderTest.cs (hunk 2 above)
Comment (iiSaLMaN) [reply osu.Game.Tests/Skins/LegacySkinDecoderTest.cs]: As told in the description of that commit, checked differently with `TestSceneSkinConfigurationLookup.TestEmptyComboColours(false)`

Comment (peppy) [reply osu.Game.Tests/Skins/LegacySkinDecoderTest.cs]: no reason to not leave the test here?

Comment (iiSaLMaN) [reply osu.Game.Tests/Skins/LegacySkinDecoderTest.cs]: it'll fail as the combo colours are gonna be empty? unless it's fine for it to be there

File: osu.Game/Skinning/LegacySkin.cs (hunk 4 above)
Comment (iiSaLMaN) [reply osu.Game/Skinning/LegacySkin.cs]: I fail to understand what you mean, care to explain further?

File: osu.Game.Tests/Skins/LegacySkinDecoderTest.cs (hunk 2 above)
Comment (peppy) [reply osu.Game.Tests/Skins/LegacySkinDecoderTest.cs]: i really don't want to have to spell it out, but it seems i have to:

update the test asserts to match the *expected* output.

File: osu.Game/Skinning/LegacySkin.cs (hunk 4 above)
Comment (peppy) [reply osu.Game/Skinning/LegacySkin.cs]: `skin.Configuration.ComboColours` is going to return different results to `skin.GetConfig(ComboColours)`

Comment (iiSaLMaN) [reply osu.Game/Skinning/LegacySkin.cs]: I'm not sure if this is causing issues though? if the combo colours retrieved is null (in DHO), the white colour will be used. what I'm trying to say is that the correct way to use this is by using a `SkinProvidingContainer`.

Comment (bdach) [reply osu.Game/Skinning/LegacySkin.cs]: Having two ways of accessing the same value (or at least what *looks like* it should be the same value, as this does) return two different results is bug-prone by default and a violation of encapsulation. We don't want to have to remember in the future which way of accessing it is *right*. There should be only one way of accessing a particular value.

Comment (iiSaLMaN) [reply osu.Game/Skinning/LegacySkin.cs]: are you talking about `skin.Configuration.ComboColours` returning empty list and `GetConfig(ComboColours` returning null?

File: osu.Game/Beatmaps/Formats/IHasComboColours.cs [hunk 5]
```diff
@@ -8,6 +8,19 @@ namespace osu.Game.Beatmaps.Formats
...
+
+        /// <summary>
+        /// Adds combo colours to the list.
+        /// </summary>
+        void AddComboColours(params Color4[] colours);
+
+        /// <summary>
+        /// Clear current combo colours from the list.
+        /// </summary>
+        void ClearComboColours();
```
Comment (peppy) [inline osu.Game/Beatmaps/Formats/IHasComboColours.cs]: when would we every want to clear combo colours?

File: osu.Game/Beatmaps/Formats/IHasComboColours.cs [hunk 6]
```diff
@@ -8,6 +8,19 @@ namespace osu.Game.Beatmaps.Formats
...
-        List<Color4> ComboColours { get; set; }
+        /// <summary>
+        /// Retrieves the list of combo colours for presentation only.
+        /// </summary>
+        IReadOnlyList<Color4> ComboColours { get; set; }
+
+        /// <summary>
+        /// Adds combo colours to the list.
+        /// </summary>
+        void AddComboColours(params Color4[] colours);
```
Comment (peppy) [inline osu.Game/Beatmaps/Formats/IHasComboColours.cs]: why was this added? i'm looking for reason for the added complexity in this PR.

File: osu.Game/Skinning/SkinConfiguration.cs [hunk 7]
```diff
@@ -14,7 +16,39 @@ public class SkinConfiguration : IHasComboColours, IHasCustomColours
...
+
+                if (AllowDefaultComboColoursFallback)
+                    return DefaultComboColours;
+
+                return null;
+            }
+            set => comboColours = value.ToList();
+        }
+
+        public void AddComboColours(params Color4[] colours) => colours.ForEach(c => comboColours.Add(c));
```
Comment (peppy) [inline osu.Game/Skinning/SkinConfiguration.cs]: if this is necessary, `AddRange` like before?

**Review by peppy — CHANGES_REQUESTED** [review]

File: osu.Game/Beatmaps/Formats/IHasComboColours.cs (hunk 6 above)
Comment (iiSaLMaN) [reply osu.Game/Beatmaps/Formats/IHasComboColours.cs]: To allow passing default combo colours when trying to read via `ComboColours`, and avoid having 2 different values like this PR used to be: (`GetConfig(ComboColours)`, `SkinConfiguration.ComboColours`).

File: osu.Game/Beatmaps/Formats/IHasComboColours.cs (hunk 5 above)
Comment (iiSaLMaN) [reply osu.Game/Beatmaps/Formats/IHasComboColours.cs]: Wasn't sure about leaving an add function without remove/clear function.., will remove.

File: osu.Game/Skinning/SkinConfiguration.cs (hunk 7 above)
Comment (iiSaLMaN) [reply osu.Game/Skinning/SkinConfiguration.cs]: Right.. 🤦‍♂️

File: osu.Game.Tests/Skins/LegacySkinDecoderTest.cs [hunk 8]
```diff
@@ -15,14 +15,24 @@ public class LegacySkinDecoderTest
     {
         [TestCase(true)]
         [TestCase(false)]
-        public void TestDecodeSkinColours(bool hasColours)
+        [TestCase(false, false)]
+        public void TestDecodeSkinColours(bool hasColours, bool canFallback = true)
```
Comment (smoogipoo) [inline osu.Game.Tests/Skins/LegacySkinDecoderTest.cs]: Please split out into multiple tests (at least the `canFallback` part). Each method should test at most one case.

**Review by smoogipoo — CHANGES_REQUESTED** [review]
This doesn't seem like a bad change.

File: osu.Game.Tests/Skins/LegacySkinDecoderTest.cs (hunk 8 above)
Comment (iiSaLMaN) [reply osu.Game.Tests/Skins/LegacySkinDecoderTest.cs]: Does this mean to not always use `TestCase(val, val)` in future changes?

Comment (smoogipoo) [reply osu.Game.Tests/Skins/LegacySkinDecoderTest.cs]: In general, yes, they should be avoided. The typical use case is having expected values and such, e.g.:

`TestMethod(bool arg1, bool expectedResult)`

or similar. Having them as args that affect the test case increases the human parsing complexity a lot.

**Comment by peppy** [comment]
This is missing a test for the case you're trying to resolve (ie. there is no beatmap test).

**Comment by iiSaLMaN** [comment]
Wrote a beatmap test for `LegacyBeatmapSkin` that retrieves combo colours of the `Player`'s `BeatmapSkinProvidingContainer` to assert with, a bit complicated but hopefully good to go.

**Review by peppy — APPROVED** [review]

**Comment by peppy** [comment]
@iiSaLMaN still required a considerable amount of tidy-up. please check my commits, especially https://github.com/ppy/osu/pull/6449/commits/783159f7a06de9438552f0de2b9654662b78b6ac and https://github.com/ppy/osu/pull/6449/commits/145ac8e0b547c3f8472f4a0344dec7b6e3a4730a.
